cpp/src/arrow/compute/exec/project_node.cc

```diff
@@ -88,10 +88,13 @@ class ProjectNode : public MapNode {
     return ExecBatch{std::move(values), target.length};
   }
 
-  void InputReceived(ExecNode* input, ExecBatch batch) override {
+  void InputReceived(ExecNode* input, std::function<Result<ExecBatch>()> task) override {
     DCHECK_EQ(input, inputs_[0]);
-    auto func = [this](ExecBatch batch) { return DoProject(std::move(batch)); };
-    this->SubmitTask(std::move(func), std::move(batch));
+    auto func = [this, task]() -> Result<ExecBatch> {
+      ARROW_ASSIGN_OR_RAISE(auto batch, task());
+      return DoProject(std::move(batch));
+    };
+    this->SubmitTask(std::move(func));
```

Comment:
```suggestion
    this->SubmitTask(std::move(func));
```
this->outputs_[0]->InputReceived(this, std::move(func));

Reply:
That's right, that change will happen when in the PRs related to the scheduler x3